File: napa/PersistentStack.m
//
//  PersistentStack.m
//  napa
//
//  Created by Remy JOURDE on 29/06/2014.
//  Copyright (c) 2014 Remy Jourde. All rights reserved.
//

#import "PersistentStack.h"

@interface PersistentStack ()

@property (nonatomic,strong,readwrite) NSManagedObjectContext* managedObjectContext;

@end

@implementation PersistentStack

- (id)init
{
    self = [super init];
    if (self) {
        [self setupManagedObjectContext];
    }
    return self;
}

- (void)setupManagedObjectContext
{
    self.managedObjectContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
    self.managedObjectContext.persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:self.managedObjectModel];
    NSError* error;
    [self.managedObjectContext.persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:self.storeURL options:nil error:&error];
    if (error) {
        NSLog(@"error: %@", error);
    }
    self.managedObjectContext.undoManager = [[NSUndoManager alloc] init];
}

- (NSManagedObjectModel*)managedObjectModel
{
    return [[NSManagedObjectModel alloc] initWithContentsOfURL:self.modelURL];
}

- (NSURL*)storeURL
{
    return [[self applicationDocumentsDirectory] URLByAppendingPathComponent:@"napa_1.2.sqlite"];
}

- (NSURL*)modelURL
{
    return [[NSBundle mainBundle] URLForResource:@"napa" withExtension:@"momd"];
}

#pragma mark - Application's Documents directory

// Returns the URL to the application's Documents directory.
- (NSURL *)applicationDocumentsDirectory
{
    return [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
}

@end

File: napa/PersistentStack.swift
//
//  PersistentStack.swift
//  napa
//
//  Created by Remy JOURDE on 29/09/2014.
//  Copyright (c) 2014 Remy Jourde. All rights reserved.
//

import CoreData

class PersistentStack: NSObject {
    
    var managedObjectModel : NSManagedObjectModel?
    var persistentStoreCoordinator: NSPersistentStoreCoordinator?
    var managedObjectContext: NSManagedObjectContext?
    
    override init() {
        super.init()
        
        if let modelURL = NSBundle.mainBundle().URLForResource("napa", withExtension: "momd") {
            if let managedObjectModel = NSManagedObjectModel(contentsOfURL: modelURL) {
                persistentStoreCoordinator = NSPersistentStoreCoordinator(managedObjectModel: managedObjectModel)
                
                var error: NSError?
                
                let storeURL = self.applicationDocumentsDirectory().URLByAppendingPathComponent("napa_1.5.sqlite")
                
                persistentStoreCoordinator!.addPersistentStoreWithType(NSSQLiteStoreType, configuration: nil, URL: storeURL, options: nil, error: &error)
                
                managedObjectContext = NSManagedObjectContext(concurrencyType: NSManagedObjectContextConcurrencyType.MainQueueConcurrencyType)
                managedObjectContext!.persistentStoreCoordinator = persistentStoreCoordinator
                managedObjectContext!.undoManager = NSUndoManager()
            }
        }
    }
    
    func saveContext() {
        var error: NSError?
        
        if let managedObjectContext = self.managedObjectContext {
            if managedObjectContext.hasChanges && !managedObjectContext.save(&error) {
                // Replace this implementation with code to handle the error appropriately.
                // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
                println("error saving context: \(error)")
                abort();
            }
        }
    }
    
    // Mark: - Application's Documents directory
    
    // Returns the URL to the application's Documents directory.
    func applicationDocumentsDirectory() -> NSURL {
        let directory = NSSearchPathDirectory.DocumentDirectory
        let domains = NSSearchPathDomainMask.UserDomainMask
        
        return NSFileManager.defaultManager().URLsForDirectory(directory, inDomains: domains).last as NSURL
    }
}
